Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp

```diff
@@ -0,0 +1,588 @@
+    case CSSUnitType::CSS_ANCHOR:
+    case CSSUnitType::CSS_ATTR:
+    case CSSUnitType::CSS_CALC:
+    case CSSUnitType::CSS_CALC_PERCENTAGE_WITH_LENGTH:
+    case CSSUnitType::CSS_CALC_PERCENTAGE_WITH_NUMBER:
+    case CSSUnitType::CSS_DIMENSION:
+    case CSSUnitType::CSS_FONT_FAMILY:
+    case CSSUnitType::CSS_IDENT:
+    case CSSUnitType::CSS_PROPERTY_ID:
+    case CSSUnitType::CSS_QUIRKY_EM:
+    case CSSUnitType::CSS_RGBCOLOR:
+    case CSSUnitType::CSS_STRING:
+    case CSSUnitType::CSS_UNKNOWN:
+    case CSSUnitType::CSS_UNRESOLVED_COLOR:
+    case CSSUnitType::CSS_URI:
+    case CSSUnitType::CSS_VALUE_ID:
+    case CSSUnitType::CustomIdent:
+        break;
+    }
+
+    return 64;
+}
+
+static unsigned sortPriority(const Child& child)
+{
+    // https://drafts.csswg.org/css-values-4/#sort-a-calculations-children
+
+    return WTF::switchOn(child,
+        []<Numeric T>(const T& root) -> unsigned {
+            return sortPriority(toCSSUnit(root));
+        },
+        [](const auto&) -> unsigned {
+            return 65; // NOTE: 65 is greater than any numeric unit type, even the error case.
+        }
+    );
+}
+
+struct ChildRep {
+    // Offset in the operations `children` vector.
+    size_t index;
+
+    // Value used to order children during sort, based on unit.
+    unsigned sortPriority;
+};
+
+static Vector<ChildRep, 16> generateSortedChildrenMap(const Children& children)
+{
+    Vector<ChildRep, 16> sortedChildrenMap;
+    sortedChildrenMap.reserveInitialCapacity(children.size());
+
+    for (size_t i = 0; i < children.size(); ++i)
+        sortedChildrenMap.append(ChildRep { .index = i, .sortPriority = sortPriority(children[i]) });
+
+    std::stable_sort(sortedChildrenMap.begin(), sortedChildrenMap.end(), [](const auto& first, const auto& second) -> bool {
+        return first.sortPriority < second.sortPriority;
+    });
+
+    return sortedChildrenMap;
+}
+
+// MARK: Math Function
+// https://drafts.csswg.org/css-values-4/#serialize-a-math-function
+
+static double clampValue(double value, ValueRange range)
+{
+    value = std::isnan(value) ? 0 : value;
+    return range == ValueRange::NonNegative && value < 0 ? 0 : value;
+}
+
+void serializeMathFunction(StringBuilder& builder, const Child& fn, SerializationState& state)
+{
+    WTF::switchOn(fn, [&builder, &state](const auto& root) { serializeMathFunction(builder, root, state); });
+}
+
+template<Numeric Op> void serializeMathFunction(StringBuilder& builder, const Op& fn, SerializationState& state)
+{
+    // 1. If the root of the calculation tree fn represents is a numeric value (number, percentage, or dimension), and the serialization being produced is of a computed value or later, then clamp the value to the range allowed for its context (if necessary), then serialize the value as normal and return the result.
+
+    if (state.stage == Stage::Computed) {
+        auto clampedFn = makeChildWithValueBasedOn(clampValue(fn.value, state.range), fn);
```

Comment:
nit: no abbreviations

Reply:
`fn` is what the spec uses, so keeping it consistent was helpful for reading comprehension.